lib/internal/process/esm_loader.js

```diff
@@ -31,31 +39,96 @@ exports.importModuleDynamicallyCallback = async function(wrap, specifier) {
   throw new ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING();
 };
 
-let ESMLoader = new Loader();
-exports.ESMLoader = ESMLoader;
-
-exports.initializeLoader = initializeLoader;
-async function initializeLoader() {
-  const { getOptionValue } = require('internal/options');
-  const userLoader = getOptionValue('--experimental-loader');
-  if (!userLoader)
-    return;
-  let cwd;
-  try {
-    cwd = process.cwd() + '/';
-  } catch {
-    cwd = 'file:///';
+let processLoader;
+exports.getLoader = () => {
+  if (!processLoader) {
+    processLoader = (async () => {
+      const loader = new Loader();
+
+      const { getOptionValue } = require('internal/options');
+      const userLoaders = getOptionValue('--experimental-loader');
+
+      if (userLoaders.length > 0) {
+        const { emitExperimentalWarning } = require('internal/util');
+        emitExperimentalWarning('--experimental-loader');
+
+        let cursor = ObjectFreeze({
+          __proto__: null,
+          resolve: loader._resolve,
+          getFormat: loader._getFormat,
+          getSource: loader._getSource,
+          transformSource: loader._transformSource,
+        });
+
+        for (let i = 0; i < userLoaders.length; i += 1) {
+          const {
+            resolve,
+            getFormat,
+            getSource,
+            transformSource,
+            getGlobalPreloadCode
+          } = await loader.importLoader(userLoaders[i]);
+
+          if (resolve !== undefined) {
+            validateFunction(resolve, 'resolve');
+          }
+          if (getFormat !== undefined) {
+            validateFunction(getFormat, 'getFormat');
+          }
+          if (getSource !== undefined) {
+            validateFunction(getSource, 'getSource');
+          }
+          if (transformSource !== undefined) {
+            validateFunction(transformSource, 'transformSource');
+          }
+
+          if (getGlobalPreloadCode !== undefined) {
+            validateFunction(getGlobalPreloadCode, 'getGlobalPreloadCode');
+            const globalPreloadCode = getGlobalPreloadCode();
+            validateString(globalPreloadCode, 'globalPreloadCode');
+            ArrayPrototypePush(loader.globalPreloadCode, globalPreloadCode);
+          }
+
+          const parentLoader = cursor;
+          cursor = ObjectFreeze({
+            __proto__: null,
+            resolve: resolve ? async (...args) => {
+              const result = await resolve(...args, parentLoader);
+              if (result === null) {
```

Comment:
Could we treat `null` and `undefined` as equals here? I personally prefer the code style of `return` with no argument over `return null` to indicate deferral.

I don't believe there is any need for `undefined` to be treated as a value return for any hook so this should be possible.

Reply:
I originally had both, but I think forcing it to be explicit is better. Only using null caught a bug at least once while I was testing this so I assume it will help others as well.

Reply:
I agree explicit is better, but `null` isn't exactly the most liked "object" out there. The other option is to require a call to the parent and have a validation error for any nullish return, which perhaps makes the intent even more clear if we want to be explicit.

Reply:
well i'm not strongly against including `undefined` again, maybe a third human can chime in?

Reply:
to clarify, if you *return* a (nullish) value it implicitly delegates? why would it delegate rather than fail?

Reply:
@bmeck It's just for convenience, you can see how much it is used in the diff.

Reply:
No strong opinions but +0.5 on excluding `undefined` (only accepting `null`). I can see scenarios where code may "forget" to handle branches exhaustively and accidentally delegate. But it's hard to tell if that'd be a problem in practice.

Reply:
I think the implicit delegation might be more the source of the problem than `undefined` in particular. I think the errors would likely have been clearer if no implicit behavior was done. Is the convenience/extra learning worth this? I think excluding `undefined` while performing default behavior isn't in line with Node, TC39, or WebIDL conventions. I do see the "please delegate if a specific return value is issued" as useful, but something about this function really is making it hard for me to grok unclear what right now.

Reply:
@bmeck yeah i only chose it because functions implicitly return undefined. maybe we can land with only null and in the future it can be loosened to allow undefined if needed?

Reply:
@devsnek it seems fine to land as stuff is still experimental and I see the utility of it, just the API feels unlike any other API we have